File: v2/src/renderer/src/utils/uri.ts
export interface ParsedUri {
  protocol: string;
  host: string;
  port: number | null;
  database: string | null;
}

export function parseMongoUri(uri: string): ParsedUri | null {
  try {
    const parsed = new URL(uri);
    if (!/^mongodb(\+srv)?:$/.test(parsed.protocol)) return null;
    return {
      protocol: parsed.protocol.replace(':', ''),
      host: parsed.hostname,
      port: parsed.port ? Number(parsed.port) : null,
      database: parsed.pathname.replace(/^\//, '') || null,
    };
  } catch {
    return null;
  }
}

export function formatUriPreview(uri: string): string {
  const p = parseMongoUri(uri);
  if (!p) return uri;
  const port = p.port ? `:${p.port}` : '';
  const db = p.database ? `/${p.database}` : '';
  return `${p.protocol}://${p.host}${port}${db}`;
}

File: v3/src/main/kotlin/io/mex/util/Uri.kt
package io.mex.util

import java.net.URI

data class ParsedUri(
    val protocol: String,
    val host: String,
    val port: Int?,
    val database: String?,
)

fun parseMongoUri(uri: String): ParsedUri? = runCatching {
    val parsed = URI(uri)
    val scheme = parsed.scheme ?: return null
    if (scheme != "mongodb" && scheme != "mongodb+srv") return null
    ParsedUri(
        protocol = scheme,
        host = parsed.host ?: "",
        port = parsed.port.takeIf { it >= 0 },
        database = parsed.path?.removePrefix("/")?.takeIf { it.isNotEmpty() },
    )
}.getOrNull()

fun formatUriPreview(uri: String): String {
    // Multi-host URIs (host1,host2) defeat java.net.URI, so the fallback must
    // still strip credentials rather than echo the raw URI.
    val p = parseMongoUri(uri) ?: return redactCredentials(uri)
    val port = p.port?.let { ":$it" } ?: ""
    val db = p.database?.let { "/$it" } ?: ""
    return "${p.protocol}://${p.host}$port$db"
}

fun redactCredentials(uri: String): String =
    Regex("(mongodb(\\+srv)?://)([^@/]+@)").replace(uri, "$1***@")
